File: ckgame-java/src/main/java/com/ckgame/military/Army.java
package com.ckgame.military;

import java.util.ArrayList;
import java.util.List;

public final class Army {
    public int id;
    public int owner;
    public int commander;
    public String name;
    public int location;
    public ArmyStatus status = ArmyStatus.IDLE;
    public final List<UnitStack> stacks = new ArrayList<>();
    public final List<Integer> path = new ArrayList<>();
    public int supply = 100;
    public int morale = 100;

    public Army(int id, int owner, String name, int location, int commander) {
        this.id = id;
        this.owner = owner;
        this.name = name;
        this.location = location;
        this.commander = commander;
    }

    public boolean isActive() {
        return status != ArmyStatus.DISBANDED && totalMen() > 0;
    }

    public int totalMen() {
        int total = 0;
        for (UnitStack stack : stacks) {
            total += Math.max(0, stack.men);
        }
        return total;
    }

    public void addMen(UnitType unitType, int men) {
        if (unitType == null || men <= 0) {
            return;
        }
        for (UnitStack stack : stacks) {
            if (stack.unitType == unitType) {
                stack.men += men;
                stack.maxMen += men;
                return;
            }
        }
        stacks.add(new UnitStack(unitType, men, men));
    }

    public double monthlyMaintenance() {
        double total = 0.0;
        for (UnitStack stack : stacks) {
            total += stack.men * stack.unitType.maintenance();
        }
        return total;
    }

    public void setPath(List<Integer> newPath) {
        path.clear();
        if (newPath == null || newPath.isEmpty()) {
            status = ArmyStatus.IDLE;
            return;
        }
        path.addAll(newPath);
        if (!path.isEmpty() && path.get(0) == location) {
            path.remove(0);
        }
        status = path.isEmpty() ? ArmyStatus.IDLE : ArmyStatus.MOVING;
    }

    public void advanceMove(double moveChance) {
        if ((status != ArmyStatus.MOVING && status != ArmyStatus.RETREATING) || path.isEmpty()) {
            return;
        }
        if (moveChance < 1.0 && Math.random() > Math.max(0.0, moveChance)) {
            return;
        }
        location = path.remove(0);
        if (status == ArmyStatus.RETREATING && !path.isEmpty()) {
            return;
        }
        status = path.isEmpty() ? ArmyStatus.IDLE : ArmyStatus.MOVING;
    }

    public void applySupplyTick(boolean inFriendlyCounty, boolean winter) {
        int delta = inFriendlyCounty ? 5 : -8;
        if (winter) {
            delta -= 4;
        }
        supply = Math.max(0, Math.min(150, supply + delta));
        if (supply < 30) {
            morale = Math.max(10, morale - 3);
        } else if (supply > 80) {
            morale = Math.min(100, morale + 1);
        }
        if (totalMen() <= 0) {
            status = ArmyStatus.DISBANDED;
        }
    }
}

File: ckgame-ts/src/military/Army.ts
import { ArmyStatus } from './ArmyStatus.js';
import { UnitType, unitMaintenanceValue } from './UnitType.js';
import { UnitStack } from './UnitStack.js';

/**
 * An army with stacks, path, supply, and morale.
 */
export class Army {
  public id: number;
  public owner: number;
  public commander: number;
  public name: string;
  public location: number;
  public status: ArmyStatus = ArmyStatus.IDLE;
  public stacks: UnitStack[] = [];
  public path: number[] = [];
  public supply: number = 100;
  public morale: number = 100;

  constructor(id: number, owner: number, name: string, location: number, commander: number) {
    this.id = id;
    this.owner = owner;
    this.name = name;
    this.location = location;
    this.commander = commander;
  }

  /** Whether the army is still active (not disbanded and has men). */
  isActive(): boolean {
    return this.status !== ArmyStatus.DISBANDED && this.totalMen() > 0;
  }

  /** Total men across all stacks. */
  totalMen(): number {
    let total = 0;
    for (const stack of this.stacks) {
      total += Math.max(0, stack.men);
    }
    return total;
  }

  /** Add men of a given unit type. */
  addMen(unitType: UnitType, men: number): void {
    if (men <= 0) return;
    for (const stack of this.stacks) {
      if (stack.unitType === unitType) {
        stack.men += men;
        stack.maxMen += men;
        return;
      }
    }
    this.stacks.push(new UnitStack(unitType, men, men));
  }

  /** Monthly maintenance cost. */
  monthlyMaintenance(): number {
    let total = 0;
    for (const stack of this.stacks) {
      total += stack.men * unitMaintenanceValue(stack.unitType);
    }
    return total;
  }

  /** Set a movement path. */
  setPath(newPath: number[]): void {
    this.path = [];
    if (!newPath || newPath.length === 0) {
      this.status = ArmyStatus.IDLE;
      return;
    }
    this.path = [...newPath];
    // Remove current location from start of path
    if (this.path.length > 0 && this.path[0] === this.location) {
      this.path.shift();
    }
    this.status = this.path.length === 0 ? ArmyStatus.IDLE : ArmyStatus.MOVING;
  }

  /** Advance along path by one step. */
  advanceMove(moveChance: number): void {
    if ((this.status !== ArmyStatus.MOVING && this.status !== ArmyStatus.RETREATING) || this.path.length === 0) {
      return;
    }
    if (moveChance < 1.0 && Math.random() > Math.max(0.0, moveChance)) {
      return;
    }
    this.location = this.path.shift()!;
    if (this.status === ArmyStatus.RETREATING && this.path.length > 0) {
      return;
    }
    this.status = this.path.length === 0 ? ArmyStatus.IDLE : ArmyStatus.MOVING;
  }

  /** Apply a supply tick. */
  applySupplyTick(inFriendlyCounty: boolean, winter: boolean): void {
    let delta = inFriendlyCounty ? 5 : -8;
    if (winter) delta -= 4;
    this.supply = Math.max(0, Math.min(150, this.supply + delta));
    if (this.supply < 30) {
      this.morale = Math.max(10, this.morale - 3);
    } else if (this.supply > 80) {
      this.morale = Math.min(100, this.morale + 1);
    }
    if (this.totalMen() <= 0) {
      this.status = ArmyStatus.DISBANDED;
    }
  }
}
